File: maxTotalDamage.go
import (
    "sort"
)

func maximumTotalDamage(power []int) int64 {
    count := map[int]int{}
    for _, v := range power {
        count[v]++
    }

    unique := make([]int, 0, len(count))
    for k := range count {
        unique = append(unique, k)
    }
    sort.Ints(unique)

    n := len(unique)
    dp := make([][4]int64, n+1)

    for i := 1; i <= n; i++ {
        val := unique[i-1]
        total := int64(val * count[val])
        // binary search for last index allowed
        prev := sort.SearchInts(unique, val-2)
        for j := 0; j < 4; j++ {
            dp[i][j] = dp[i-1][j]
            if j > 0 && dp[i-1][j-1] > dp[i][j] {
                dp[i][j] = dp[i-1][j-1]
            }
            if dp[prev][j]+total > dp[i][j] {
                dp[i][j] = dp[prev][j] + total
            }
        }
    }
    ans := int64(0)
    for j := 0; j < 4; j++ {
        if dp[n][j] > ans {
            ans = dp[n][j]
        }
    }
    return ans
}

File: maxTotalDamage.py
from collections import Counter
from bisect import bisect_right

class Solution:
    def maximumTotalDamage(self, power):
        count = Counter(power)
        unique = sorted(count)
        n = len(unique)
        dp = [[0]*4 for _ in range(n+1)]  # dp[i][j] = max dmg using first i spells, skipping j unique spells

        for i in range(1, n+1):
            val = unique[i-1]
            total = val * count[val]
            # find prev index we can take (no conflict within 2 range)
            prev = bisect_right(unique, val - 3)
            for j in range(4):
                # skip this spell
                dp[i][j] = max(dp[i][j], dp[i-1][j])
                if j > 0:
                    dp[i][j] = max(dp[i][j], dp[i-1][j-1])
                # take this spell
                dp[i][j] = max(dp[i][j], dp[prev][j] + total)
        return max(dp[n])
